File: dashboard/routes/tasks.py
"""Dispatching work to machines, and reading back what happened."""

import csv
import io
import os
import signal
import threading
from flask import Blueprint, Response, jsonify, render_template, request

import fleet
from machines import get_computer_by_name, list_computers
from security import require_token
from settings import PAGE_SIZE
from tasks import build_analytics, compute_duration_seconds, create_task_row, dispatch_task, get_task_row, run_task_worker, update_task_row

from db import get_db


bp = Blueprint("tasks", __name__)


@bp.route("/partials/tasks")
def partial_tasks():
    desktop = (request.args.get("desktop") or "").strip()
    status = (request.args.get("status") or "").strip().upper()
    try:
        page = max(1, int(request.args.get("page") or 1))
    except ValueError:
        page = 1

    where, params = [], []
    if desktop:
        where.append("desktop = ?")
        params.append(desktop)
    if status == "ACTIVE":
        where.append("status IN ('PENDING', 'RUNNING')")
    elif status in ("COMPLETED", "FAILED", "CANCELLED"):
        where.append("status = ?")
        params.append(status)
    clause = (" WHERE " + " AND ".join(where)) if where else ""

    db = get_db()
    total = db.execute(f"SELECT COUNT(*) c FROM tasks{clause}", params).fetchone()["c"]
    pages = max(1, (total + PAGE_SIZE - 1) // PAGE_SIZE)
    page = min(page, pages)
    rows = [dict(r) for r in db.execute(
        f"SELECT * FROM tasks{clause} ORDER BY id DESC LIMIT ? OFFSET ?",
        (*params, PAGE_SIZE, (page - 1) * PAGE_SIZE),
    ).fetchall()]
    for r in rows:
        r["duration_seconds"] = compute_duration_seconds(r)

    return render_template(
        "_tasks.html",
        tasks=rows,
        names=[c["name"] for c in list_computers()],
        sel_desktop=desktop,
        sel_status=status,
        page=page, pages=pages, total=total,
    )


@bp.route("/partials/analytics")
def partial_analytics():
    return render_template("_analytics.html", analytics=build_analytics())


@bp.route("/api/v1/analytics")
def api_analytics():
    return jsonify({"ok": True, "data": build_analytics(), "error": None})


@bp.route("/api/v1/tasks", methods=["GET"])
def api_tasks_list():
    db = get_db()
    rows = [dict(r) for r in db.execute("SELECT * FROM tasks ORDER BY id DESC LIMIT 100").fetchall()]
    for r in rows:
        r["duration_seconds"] = compute_duration_seconds(r)
    return jsonify({"ok": True, "data": rows, "error": None})


@bp.route("/api/v1/tasks/export.csv")
def api_tasks_export():
    db = get_db()
    rows = db.execute("SELECT * FROM tasks ORDER BY id DESC").fetchall()
    buf = io.StringIO()
    writer = csv.writer(buf)
    writer.writerow(["id", "desktop", "description", "status", "result_text", "cost_usd",
                     "duration_seconds", "error", "created_at", "updated_at"])
    for r in rows:
        d = dict(r)
        writer.writerow([d["id"], d["desktop"], d["description"], d["status"],
                         d.get("result_text") or "", d.get("cost_usd") or "",
                         compute_duration_seconds(d) or "", d.get("error") or "",
                         d["created_at"], d["updated_at"]])
    return Response(buf.getvalue(), mimetype="text/csv",
                    headers={"Content-Disposition": "attachment; filename=deskswarm-tasks.csv"})


@bp.route("/api/v1/tasks/<int:task_id>", methods=["GET"])
def api_task_detail(task_id: int):
    row = get_task_row(task_id)
    if not row:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    row["duration_seconds"] = compute_duration_seconds(row)
    return jsonify({"ok": True, "data": row, "error": None})


@bp.route("/api/v1/tasks", methods=["POST"])
@require_token
def api_tasks_create():
    payload = request.get_json(silent=True) or {}
    description = (payload.get("description") or "").strip()
    target = payload.get("desktop") or "all"

    if not description:
        return jsonify({"ok": False, "data": None, "error": "description is required"}), 400
    try:
        created_ids = dispatch_task(target, description)
    except ValueError as exc:
        return jsonify({"ok": False, "data": None, "error": str(exc)}), 400

    return jsonify({"ok": True, "data": {"task_ids": created_ids}, "error": None}), 201


@bp.route("/api/v1/tasks/<int:task_id>/cancel", methods=["POST"])
@require_token
def api_task_cancel(task_id: int):
    row = get_task_row(task_id)
    if not row:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    if row["status"] not in ("PENDING", "RUNNING"):
        return jsonify({"ok": False, "data": None, "error": f"task is {row['status']}, cannot cancel"}), 400
    update_task_row(task_id, status="CANCELLED", current_action=None, error="cancelled by user")
    if row.get("pid"):
        try:
            os.kill(row["pid"], signal.SIGTERM)
        except ProcessLookupError:
            pass
    return jsonify({"ok": True, "data": {"id": task_id, "status": "CANCELLED"}, "error": None})


@bp.route("/api/v1/tasks/<int:task_id>/retry", methods=["POST"])
@require_token
def api_task_retry(task_id: int):
    row = get_task_row(task_id)
    if not row:
        return jsonify({"ok": False, "data": None, "error": "not found"}), 404
    comp = get_computer_by_name(row["desktop"])
    if not comp:
        return jsonify({"ok": False, "data": None, "error": f"computer '{row['desktop']}' no longer exists"}), 400

    new_id = create_task_row(comp["name"], row["description"])
    threading.Thread(
        target=run_task_worker,
        args=(new_id, fleet.bridge_container_name(comp["slug"]), 8000, row["description"]),
        daemon=True,
    ).start()
    return jsonify({"ok": True, "data": {"task_id": new_id}, "error": None}), 201

File: dashboard-ts/src/routes/tasks.ts
/** Dispatching work to machines, and reading back what happened. */

import { Hono } from "hono";
import * as fleet from "./../fleet";
import { all, one } from "./../db";
import { getComputerByName, listComputers } from "./../machines";
import {
  buildAnalytics,
  computeDurationSeconds,
  createTaskRow,
  dispatchTask,
  getTaskRow,
  runTaskWorker,
  updateTaskRow,
  ValidationError,
  type TaskRow,
} from "./../tasks";
import { render } from "./../templates";
import { requireToken } from "./../security";
import { PAGE_SIZE } from "./../settings";
import { fail, intParam, jsonBody, notFound, ok, type Env } from "./../http";

export const tasks = new Hono<Env>();

/** The filter the tasks table and its partial share. */
function queryTasks(desktop: string, status: string, page: number) {
  const where: string[] = [];
  const params: any[] = [];
  if (desktop) {
    where.push("desktop = ?");
    params.push(desktop);
  }
  if (status === "ACTIVE") {
    where.push("status IN ('PENDING', 'RUNNING')");
  } else if (["COMPLETED", "FAILED", "CANCELLED"].includes(status)) {
    where.push("status = ?");
    params.push(status);
  }
  const clause = where.length ? " WHERE " + where.join(" AND ") : "";

  const total =
    one<{ c: number }>(`SELECT COUNT(*) c FROM tasks${clause}`, ...params)?.c ?? 0;
  const pages = Math.max(1, Math.ceil(total / PAGE_SIZE));
  const wanted = Math.min(page, pages);
  const rows = all<TaskRow>(
    `SELECT * FROM tasks${clause} ORDER BY id DESC LIMIT ? OFFSET ?`,
    ...params,
    PAGE_SIZE,
    (wanted - 1) * PAGE_SIZE,
  );
  for (const r of rows) r.duration_seconds = computeDurationSeconds(r);
  return { rows, pages, page: wanted, total };
}

tasks.get("/partials/tasks", (c) => {
  const desktop = (c.req.query("desktop") ?? "").trim();
  const status = (c.req.query("status") ?? "").trim().toUpperCase();
  const asked = parseInt(c.req.query("page") ?? "1", 10);
  const { rows, pages, page, total } = queryTasks(
    desktop,
    status,
    Number.isFinite(asked) ? Math.max(1, asked) : 1,
  );
  return c.html(
    render("_tasks.html", {
      tasks: rows,
      names: listComputers().map((x) => x.name),
      sel_desktop: desktop,
      sel_status: status,
      page,
      pages,
      total,
    }),
  );
});

tasks.get("/partials/analytics", (c) =>
  c.html(render("_analytics.html", { analytics: buildAnalytics() })),
);

tasks.get("/api/v1/analytics", (c) => ok(c, buildAnalytics()));

tasks.get("/api/v1/tasks", (c) => {
  const rows = all<TaskRow>("SELECT * FROM tasks ORDER BY id DESC LIMIT 100");
  for (const r of rows) r.duration_seconds = computeDurationSeconds(r);
  return ok(c, rows);
});

tasks.get("/api/v1/tasks/export.csv", (c) => {
  const rows = all<TaskRow>("SELECT * FROM tasks ORDER BY id DESC");
  const cell = (v: unknown) => {
    const s = v === null || v === undefined ? "" : String(v);
    return /[",\n\r]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
  };
  const lines = [
    ["id", "desktop", "description", "status", "result_text", "cost_usd",
     "duration_seconds", "error", "created_at", "updated_at"].join(","),
  ];
  for (const d of rows) {
    lines.push(
      [
        d.id, d.desktop, d.description, d.status, d.result_text ?? "",
        d.cost_usd ?? "", computeDurationSeconds(d) ?? "", d.error ?? "",
        d.created_at, d.updated_at,
      ].map(cell).join(","),
    );
  }
  return c.body(lines.join("\r\n") + "\r\n", 200, {
    "content-type": "text/csv",
    "content-disposition": "attachment; filename=deskswarm-tasks.csv",
  });
});

tasks.get("/api/v1/tasks/:id", (c) => {
  const id = intParam(c, "id");
  const row = id === null ? null : getTaskRow(id);
  if (!row) return notFound(c);
  row.duration_seconds = computeDurationSeconds(row);
  return ok(c, row);
});

tasks.post("/api/v1/tasks", requireToken, async (c) => {
  const payload = await jsonBody(c);
  const description = String(payload.description ?? "").trim();
  const target = String(payload.desktop || "all");
  if (!description) return fail(c, "description is required");
  try {
    return ok(c, { task_ids: dispatchTask(target, description) }, 201);
  } catch (err: any) {
    if (err instanceof ValidationError) return fail(c, String(err.message));
    throw err;
  }
});

tasks.post("/api/v1/tasks/:id/cancel", requireToken, (c) => {
  const id = intParam(c, "id");
  const row = id === null ? null : getTaskRow(id);
  if (!row) return notFound(c);
  if (!["PENDING", "RUNNING"].includes(row.status)) {
    return fail(c, `task is ${row.status}, cannot cancel`);
  }
  updateTaskRow(row.id, {
    status: "CANCELLED",
    current_action: null,
    error: "cancelled by user",
  });
  if (row.pid) {
    try {
      process.kill(row.pid, "SIGTERM");
    } catch {
      /* already gone — the row is what mattered */
    }
  }
  return ok(c, { id: row.id, status: "CANCELLED" });
});

tasks.post("/api/v1/tasks/:id/retry", requireToken, (c) => {
  const id = intParam(c, "id");
  const row = id === null ? null : getTaskRow(id);
  if (!row) return notFound(c);
  const comp = getComputerByName(row.desktop);
  if (!comp) return fail(c, `computer '${row.desktop}' no longer exists`);

  const newId = createTaskRow(comp.name, row.description);
  void runTaskWorker(newId, fleet.bridgeContainerName(comp.slug), 8000, row.description);
  return ok(c, { task_id: newId }, 201);
});
